### ISSUE TITLE
Support for multiple geometry types in one layer

### ISSUE THREAD

**Issue by arbakker** [t-rex-tileserver/t-rex#151]
Great job on t-rex, it looks very promising for the project I am currently working on.

However, for this project there is the requirement to tile a dataset with multiple geometry types into one vector tile layer.

The [MVT spec](https://github.com/mapbox/vector-tile-spec/tree/master/2.1#42-features) allows for this since the geometry type is defined on each feature and not on the layer. t-rex Seems to [require](https://github.com/t-rex-tileserver/t-rex/blob/556db7392d3127cd0b99e95a1c17278185eb1f5d/t-rex-core/src/datasource/postgis_fields.rs#L15) the geometry type for each layer to be known beforehand.

Would there be interest in support for vector tile layers with multiple geometry types in t-rex (so setting the geometry type to `GEOMETRY`)? I imagine it would be possible for t-rex to dynamically detect the feature type for each feature, after retrieving it from the datasource (correct me if I am wrong though).

Another approach could be to define multiple layers with the same name but different geometry type in the config file. These different layers then would be merged into one MVT layer. For example:

```
[[tileset.layer]]
name = "Planobject"
datasource = "dbconn"
table_name = "planobject"
geometry_field = "wkb_geometry"
geometry_type = "MULTIPOLYGON"
buffer_size = 8
simplify = true
srid = 28992

[[tileset.layer]]
name = "Planobject"
datasource = "dbconn"
table_name = "planobject"
geometry_field = "wkb_geometry"
geometry_type = "MULTILINESTRING"
buffer_size = 8
simplify = true
srid = 28992
```

**Comment by pka** [comment]
After re-reading the MVT spec (for at least the 10th time), I'm shocked that I was always under the impression that a layer can only have one geometry type. And for me, also Mapbox GL JSON styling seemed to make the same assumption. Just out of curiosity, what's the advantage in your case of mixing geometry types in a single layerer compared to multiple layers of a tileset?

I think, the natural counterpart in PostGIS would be the `GEOMETRY` type, indeed. I'll check whether supporting a mixed `GEOMETRY` type is feasable.

**Comment by flother** [comment]
> Just out of curiosity, what's the advantage in your case of mixing geometry types in a single layerer compared to multiple layers of a tileset?

This is something I'd be excited to see in the next version of T-rex too. The most obvious case that I've noticed is labelling natural features. In an ongoing project of mine, I'd like to include a single layer that contains name labels for waterfalls (points), rivers (linestrings), and lakes (polygons). Mapbox does [something similar in its `natural_labels` layer in Mapbox Streets v8](https://docs.mapbox.com/vector-tiles/reference/mapbox-streets-v8/#natural_label).

**Comment by arbakker** [comment]
> Just out of curiosity, what's the advantage in your case of mixing geometry types in a single layerer compared to multiple layers of a tileset?

The dataset (Dutch spatial plans) we are creating a tileset for is rather heterogeneous in respect to geometry types. Complex gml containg all sorts of  geometries. We can split the data in different layers by geometry type, but this would not be a meaningful subdivision of the dataset, just adding (a bit of) complexity to the solution.

Also previously we used gdal-ogr to generate the vector tiles, which is capable of multiple geometry types per layer. So we already built processing pipeline around this, but we ran into unpredictable memory usage of gdal-ogr.  I would like to be able to simply swap out gdal/ogr in our processing pipeline for trex.

**Comment by kraftto** [comment]
Any update on this? We're also working with spatial plans (in this case german) where there are many different layers stored in a DB with generic GEOMETRY.

**Comment by pka** [comment]
Implemented in 07ffc05 for PostGIS type `GEOMETRY`.

Tested with:
```sql
CREATE TABLE labels (
  id  SERIAL PRIMARY KEY,
  wkb_geometry GEOMETRY(GEOMETRY, 3857) NOT NULL,
 label TEXT
);
INSERT INTO labels (wkb_geometry,label) VALUES ('SRID=3857;POINT(1376478 7231554)', 'point');
INSERT INTO labels (wkb_geometry,label) VALUES ('SRID=3857;POLYGON((1376278 7231699,1375968 7231083,1375725 7230672,1376278 7231699))', 'poly');
```

Simplification is not applied to layers of type `GEOMETRY`.
